src/main/java/software/amazon/smithy/lsp/SmithyTextDocumentService.java

```diff
@@ -197,6 +225,97 @@ private String findToken(String path, Position p) throws IOException {
+    private Position documentStart = new Position(0, 0);
+
+    private TextEdit insertLine(String line, DocumentPreamble preamble) {
+        String trailingNewLine;
+        if (!preamble.isBlankSeparated())
+            trailingNewLine = "\n";
+        else
+            trailingNewLine = "";
+        // case 1 - there's no use block at all, so we need to insert the line directly
+        // under namespace
+        if (preamble.getUseBlockRange().getStart() == blankPosition) {
+            // case 1.a - there's no namespace - that means the document is invalid
+            // so we'll just insert the line at the beginning of the document
+            if (preamble.getNamespaceRange().getStart() == blankPosition) {
+                return new TextEdit(new Range(documentStart, documentStart), line + trailingNewLine);
+            } else {
+                Position namespaceEnd = preamble.getNamespaceRange().getEnd();
+                namespaceEnd.setCharacter(namespaceEnd.getCharacter() + 1);
+                return new TextEdit(new Range(namespaceEnd, namespaceEnd), "\n" + line + trailingNewLine);
+            }
+        } else {
+            Position useBlockEnd = preamble.getUseBlockRange().getEnd();
+            return new TextEdit(new Range(useBlockEnd, useBlockEnd), "\n" + line + trailingNewLine);
+        }
+    }
+
+    private DocumentPreamble documentPreamble(List<String> lines) {
+
+        Range namespace = new Range(blankPosition, blankPosition);
+        Range useBlock = new Range(blankPosition, blankPosition);
+        Set<String> imports = new HashSet();
+
+        // First, we detect the namespace and use block in a very ugly way
+        boolean collectUseBlock = true;
+        int lineNumber = 0;
+        for (String line : lines) {
+            if (line.trim().startsWith("namespace ")) {
+                if (namespace.getStart() == blankPosition) {
+                    namespace.setStart(new Position(lineNumber, 0));
+                    namespace.setEnd(new Position(lineNumber, line.length() - 1));
+                }
+            } else if (line.trim().startsWith("use ")) {
+                String i = line.trim().split("use ", 2)[1].trim();
+                if (useBlock.getStart() == blankPosition) {
+                    imports.add(i);
+                    useBlock.setStart(new Position(lineNumber, 0));
+                    useBlock.setEnd(new Position(lineNumber, line.length()));
+                } else if (collectUseBlock) {
+                    imports.add(i);
+                    useBlock.setEnd(new Position(lineNumber, line.length()));
+                }
+            } else if (line.isBlank()) {
+                if (collectUseBlock)
+                    useBlock.setEnd(new Position(lineNumber, line.length()));
+            } else if (line.trim().startsWith("//")) {
+
+            } else {
+                collectUseBlock = false;
+            }
+
+            lineNumber++;
+        }
```

Comment:
Given that `use` statements have strict rules about their location (they can only appear after namespace, and before any non-use statement), we only have to account for comments and newlines (which can separate parts of the `use` block)

We're doing it this way because it's more flexible - attempting to go through the parser would mean we can't add auto-imports to syntactically broken documents, which can happen if you're editing the document and it's not parseable yet.